Phili

philouail opened the pull request:
Here is the PR related to the issue #77. 

These basic functions return single values and vectors (numeric for all expect for `rowBlank` which is logical). 

This is a first draft, input is welcome, specifically on: 
- naming of functions: for example should this really be called `filteringFunctions` ? as they don't really filter but more compute measurement of precision and detection  / quality measurement of our data. I think a name reflecting that would be more accurate. `qualityFunctions` ? 
- `rowBlank` threshold 

Also, @jorainer I took your description and code for the rsd and rowRsd, it therefore needs to be removed from CompMetaboTools if you are fine with that. 

I am currently writing the vignette but wanted to check if that was what you had in mind for now. 
Also next I will implement more user friendly function that actually filter the data.

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+#'  standard deviation (i.e. coefficient of variation) of a numerical vector
+#'  or for rows of a numerical matrix, respectively.
+#'
+#' - `rowDratio` computes the D-ratio or "dispersion ratio," defined as the
```
please replace `"dispersion ratio, "` with `*dispersion ratio*,` (that will use italics font - also, the comma was inside the quotes).

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+#'  missing values in a vector or for each row of a matrix, respectively.
+#'
+#' - `rowBlank` identifies rows (i.e features) where the mean of test samples
+#'  is greater than twice the mean of blank samples. This can highlights
```
highlight instead of highlight**s**

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+#' - `rowBlank` identifies rows (i.e features) where the mean of test samples
+#'  is greater than twice the mean of blank samples. This can highlights
+#'  features that results from contamination in the solvent of the samples.
+#'   Return a `logical` vector
```
Maybe rather *Returns a `logical` vector of length equal to the number of rows of `x`.*

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+#'  matrix representing the biological samples.
+#'
+#' @param y `numeric` For `rowDratio` and `rowBlank`, a numeric matrix
+#'  representing the QC samples and blank samples, respectively.
```
maybe here *representing feature abundances in QC samples or blank samples, respectively.*

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+#' @return  See individual function description above for details.
+#'
+#' @author Philippine Louail, Johannes Rainer
+#'
```
Please add a `@notes` section specifying that, for `rsd` and `rowRsd` the feature abundances should be/are expected to be provided in natural scale and not e.g. log2 scale.

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+
+#' @export
+#' @rdname filteringFunctions
+#'
```
please remove the additional `#'` and empty line here.

jorainer replied:
Also below. I find it *cleaner* if the roxygen comments are immediately above the code, e.g.

```r
#' @rdname filterFunctions
rowRsd <- function...
```

instead of 

```r
#' @rdname filterFunctions
#'

rowRsd <- function ...
```

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,147 @@
+rowBlank <- function(x, y, na.rm = TRUE){
+    m_samples <- apply(x, 1, mean, na.rm = na.rm)
+    m_blank <- apply(y, 1, mean, na.rm = na.rm)
+    vec <- m_samples > 2 * m_blank
```
here it's enough to have

```r
m_samples > 2 * m_blank
```

otherwise (with `vec <-`) the function will not return anything (I believe?). 

Also, from the name of the function I would have expected that the function returns `TRUE` if the feature represents signal also found in blanks - i.e. the opposite what is returned now? So, basically, `rowBlank` returns `TRUE` if the signal in test samples is not at least 2 times higher than the signal in blanks, or in other words, if the feature was also found in blank samples.

philouail replied:
Yes i realised that while writing the vignette, I am now reporting TRUE for potential contaminants, it makes more sense

philouail replied:
- just waiting to make all the changes before pushing :)

jorainer replied:
the reason to drop the final `vec <- ` is also that R does not need to initialize another (then unneeded) variable and copy content in it.

jorainer commented on tests/testthat/test_function-filtering.R:
```diff
@@ -0,0 +1,26 @@
+    expect_equal(rowRsd(A), apply(A, 1, function(row) sd(row) / mean(row)))
+
+    # Test rowDratio function
+    expect_equal(as.numeric(rowDratio(A, A)), rep(1, nrow(A)))
```
please also add tests using MAD, for rsd and dratio

jorainer reviewed (changes requested):
Thanks for the great contribution @philouail ! I have some minor change requests.

philouail commented:
Thanks for the comments I fixed them, do you have an opinion on the naming thing I was talking about ?  replace `filteringFunctions` with  some kind of name that include `quality` instead ? as these functions don't filter. 

I added a section to the vignettes to describe general use of these functions.

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,142 @@
+#' @title Basic filtering functions for metabolomics
```
hm yes, now I see get your thoughts on the naming of the file. Indeed, filtering does not make much sense, maybe use "quality-assessment.R" for the file name and here `Basic quality assessment functions for metabolomics" as title?

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,142 @@
+#' When dealing with metabolomics results, it is often necessary to filter
+#' features based on certain criteria. These criteria are typically derived
+#' from statistical formulas applied to full rows of data, where each row
+#' represents a feature. The following functions provide basic filtering
```
also here then maybe better to talk about quality assessment instead. e.g. "The following functions allow to calculate basic quality assessment estimates commonly used..."

jorainer commented on R/function-filtering.R:
```diff
@@ -0,0 +1,142 @@
+#'
+#' @importFrom stats sd mad median
+#'
+#' @name filteringFunctions
```
maybe use here `@name quality-assessment` and then also below for the `@rdname` tags of all functions.

jorainer replied:
Actually, maybe `@name quality_assessment` would be better here - see comment further below in the vignette for explanation.

jorainer commented on vignettes/MetaboCoreUtils.Rmd:
```diff
@@ -546,6 +546,100 @@ Generally, injecting study samples in random order can reduce (or even avoid)
 influence of any related technical bias in the downstream analysis and is highly
 suggested to improve and assure data quality.
 
+## Filtering data: Identifying measurement error 
```
Maybe better use here a title `## Basic quality assessment and pre-filtering of metabolomics data`

jorainer commented on vignettes/MetaboCoreUtils.Rmd:
```diff
@@ -546,6 +546,100 @@ Generally, injecting study samples in random order can reduce (or even avoid)
+features based on certain criteria. These criteria are typically derived
+from statistical formulas applied to full rows of data, where each row
+represents a feature. In this tutorial, we'll explore a set of functions
+designed for filtering metabolomics data.
```
maybe change to something similar to:

...designed to calculate basic quality assessment metrics on which metabolomics data can subsequently be filtered.

jorainer commented on vignettes/MetaboCoreUtils.Rmd:
````diff
@@ -546,6 +546,100 @@ Generally, injecting study samples in random order can reduce (or even avoid)
+First, to get more information on the available function you can check the documentation 
+
+```{r}
+?filteringFunctions
````
hm, here I'm not sure if `?quality-assessment` would work? maybe then `@name quality_assessment` might be better because a `_` does not interpreted as a mathematical operation (`-` will most likely be interpreted as a minus and hence `?quality-assessment` will throw an error).

jorainer commented on vignettes/MetaboCoreUtils.Rmd:
````diff
@@ -546,6 +546,100 @@ Generally, injecting study samples in random order can reduce (or even avoid)
+print(cv_result)
+```
+
+Next, we will compute the D-ratio, a measure of dispersion, by comparing the
````
please add the reference here `[@broadhurst_guidelines_2018]` - either after "D-ratio" or "measure of dispersion"

jorainer reviewed (approved):
Great contribution @philouail ! Also very clear description of the functionality in the vignette! 

Also, I agree that "filtering" might not be the right term here - I would maybe do some re-formulations (talking about quality assessment and less of filtering) and also rename the R file accordingly.

philouail commented:
The naming system is now fixed !

@michaelwitting what do you think ? any ideas for improvement ?
